### PR TITLE
Add validator rank validation job

### LINKED ISSUE #1459: Add validator rank validation job
### Description

Currently, Lisk Service fetches all the validators from the node and computes the `rank` internally.
With `pos_getValidatorsByStake` we can fetch all the `eligible` validators by their rank. Lisk Service can use it to periodically validate the information and recompute it again, if necessary.

### Acceptance Criteria

- TODO

### Additional Information

- TODO

### PR DESCRIPTION
**Pull request by priojeetpriyom**
### What was the problem?
This PR resolves #1459

### How was it solved?
- [x] Add validator rank validation job
- [x] Add test coverage
- [x] Add support to configure jobs in all microservices
 
### How was it tested?
- [x] Local

### CODE REVIEW

File: services/blockchain-indexer/jobs/dataService/validators.js [hunk 1]
```diff
@@ -47,4 +48,19 @@ module.exports = [
 			}
 		},
 	},
+	{
+		name: 'validate.validators',
+		description: 'Validate cached validators and reload cache if necessary',
+		schedule: '0 * * * *', // Every hour
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/validators.js:54]: Can we add `interval`/`schedule` support in the config as in https://github.com/LiskHQ/lisk-service/blob/release/0.7.0/services/transaction-statistics/jobs/transactionStatistics.js#L31-L32.
```suggestion
		description: 'Validate cached validators and reload cache if necessary.',
		schedule: '4-/15 * * * *', // Every 15 minutes starting at 4 minutes past the hour
```

File: services/blockchain-indexer/shared/dataService/pos/validators.js [hunk 2]
```diff
@@ -164,6 +164,7 @@ const loadAllPosValidators = async () => {
 	}
 };
 
+// eslint-disable-next-line import/prefer-default-export
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/dataService/pos/validators.js:167]: Why do we need this?

File: services/blockchain-indexer/shared/jobs/validateValidatorsCache.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+	if (nodeValidators.length > cachedValidators.length) {
+		logger.warn(`Node has more validators than cache. node validators count:${nodeValidators.length} cached validators count:${cachedValidators.length}.`);
+		await reloadValidatorCache();
+		return;
+	}
+
+	for (let index = 0; index < nodeValidators.length; index++) {
+		if (nodeValidators[index].address !== cachedValidators[index].address) {
+			logger.warn('Mismatch found in validator rank order of node and cache. Reloading validators cache.');
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/jobs/validateValidatorsCache.js:40]:
```suggestion
			logger.warn('Incorrect validator ranking detected. Reloading the validators cache.');
```

File: services/blockchain-indexer/shared/jobs/validateValidatorsCache.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+			// eslint-disable-next-line no-await-in-loop
+			await reloadValidatorCache();
+			return;
+		}
+	}
+
+	logger.debug('Validator cache order matched with node validator ranking');
```
Comment (sameersubudhi) [inline services/blockchain-indexer/shared/jobs/validateValidatorsCache.js:47]:
```suggestion
	logger.debug('Cached validator ranks verified successfully.');
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/blockchain-indexer/shared/dataService/pos/validators.js (hunk 2 above)
Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/dataService/pos/validators.js:167]: We do not need this anymore. had added to test some way of doing unit test which did not work but forgot to remove. will remove it.

File: services/blockchain-indexer/shared/jobs/validateValidatorsCache.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+} = require('../dataService');
+
+const business = require('../dataService/business');
+
+const logger = Logger();
+
+const validateValidatorCache = async () => {
+	const validatorsByStake = await business.getPosValidatorsByStake({ limit: -1 });
+	// Create a deep copy of cached validators
+	const cachedValidators = (await getAllValidators()).slice();
```
Comment (nagdahimanshu) [inline services/blockchain-indexer/shared/jobs/validateValidatorsCache.js:30]: May we can use lodash to create deep copy of the cached validators.

Comment (priojeetpriyom) [reply services/blockchain-indexer/shared/jobs/validateValidatorsCache.js]: Applied.

File: services/blockchain-app-registry/jobs/deleteNonMetadata.js [hunk 6]
```diff
@@ -51,12 +51,13 @@ module.exports = [
 	{
 		name: 'delete.non.metadata.files',
 		description: 'Delete any non-metadata files and empty folders inside data directory.',
-		schedule: '0 0 * * *', // Every day at midnight
+		interval: config.job.deleteNonMetadataFile.interval,
+		schedule: config.job.deleteNonMetadataFile.schedule,
```
Comment (sameersubudhi) [inline services/blockchain-app-registry/jobs/deleteNonMetadata.js:55]:
```suggestion
		interval: config.job.deleteNonMetadataFiles.interval,
		schedule: config.job.deleteNonMetadataFiles.schedule,
```

File: services/blockchain-app-registry/config.js [hunk 7]
```diff
@@ -94,4 +94,16 @@ config.CHAIN_ID_PREFIX_NETWORK_MAP = Object.freeze({
...
 
+config.job = {
+	// Interval takes priority over schedule and must be greater than 0 to be valid
+	deleteNonMetadataFile: {
+		interval: process.env.DELETE_NON_METADATA_FILE_INTERVAL || 0,
+		schedule: process.env.DELETE_NON_METADATA_FILE_SCHEDULE || '0 0 * * *',
+	},
+	updateMetadata: {
+		interval: process.env.UPDATE_METADATA_INTERVAL || 0,
+		schedule: process.env.UPDATE_METADATA_SCHEDULE || '*/10 * * * *',
```
Comment (sameersubudhi) [inline services/blockchain-app-registry/config.js:105]:
```suggestion
	updateApplicationMetadata: {
		interval: process.env.UPDATE_METADATA_INTERVAL || 0,
		schedule: process.env.UPDATE_METADATA_SCHEDULE || '*/10 * * * *',
```

File: services/blockchain-app-registry/jobs/metadata.js [hunk 8]
```diff
@@ -15,13 +15,15 @@
...
+const config = require('../config');
 const { syncWithRemoteRepo } = require('../shared/utils/downloadRepository');
 
 module.exports = [
 	{
 		name: 'update.application.metadata',
 		description: 'Keep the blockchain applications metadata up-to-date',
-		schedule: '*/10 * * * *', // Every 10 minutes
+		interval: config.job.updateMetadata.interval,
+		schedule: config.job.updateMetadata.schedule,
```
Comment (sameersubudhi) [inline services/blockchain-app-registry/jobs/metadata.js:26]:
```suggestion
		interval: config.job.updateApplicationMetadata.interval,
		schedule: config.job.updateApplicationMetadata.schedule,
```

File: services/blockchain-connector/config.js [hunk 9]
```diff
@@ -102,6 +102,10 @@ config.job = {
 		interval: process.env.CACHE_CLEANUP_INTERVAL || 0,
 		schedule: process.env.CACHE_CLEANUP_SCHEDULE || '0 */12 * * *',
 	},
+	refreshPeers: {
+		interval: process.env.REFRESH_PEERS_INTERVAL || 60,
+		schedule: process.env.REFRESH_PEERS_SCHEDULE || '* * * * *',
```
Comment (sameersubudhi) [inline services/blockchain-connector/config.js:107]:
```suggestion
		schedule: process.env.REFRESH_PEERS_SCHEDULE || '',
```

File: services/blockchain-coordinator/config.js [hunk 10]
```diff
@@ -72,4 +72,12 @@ config.queue = {
 	},
 };
 
+config.job = {
+	// Interval takes priority over schedule and must be greater than 0 to be valid
+	missingBlocks: {
```
Comment (sameersubudhi) [inline services/blockchain-coordinator/config.js:77]:
```suggestion
	indexMissingBlocks: {
```

File: services/blockchain-coordinator/jobs/missingBlocks.js [hunk 11]
```diff
@@ -23,7 +24,8 @@ module.exports = [
 	{
 		name: 'index.missing.blocks',
 		description: 'Verify and update blocks indexing',
-		schedule: '*/15 * * * *', // Every 15 min
+		interval: config.job.missingBlocks.interval,
+		schedule: config.job.missingBlocks.schedule,
```
Comment (sameersubudhi) [inline services/blockchain-coordinator/jobs/missingBlocks.js:28]:
```suggestion
		interval: config.job.indexMissingBlocks.interval,
		schedule: config.job.indexMissingBlocks.schedule,
```

File: services/blockchain-indexer/jobs/dataService/blockchainAppsStats.js [hunk 12]
```diff
@@ -23,7 +25,8 @@ module.exports = [
 	{
 		name: 'reload.blockchain.apps.statistics',
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/blockchainAppsStats.js:26]:
```suggestion
		name: 'refresh.blockchain.apps.statistics',
```

File: services/blockchain-indexer/jobs/dataService/indexStatus.js [hunk 13]
```diff
@@ -26,7 +27,8 @@ module.exports = [
 	{
 		name: 'reload.live.indexing.job.count',
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/indexStatus.js:28]:
```suggestion
		name: 'refresh.live.indexing.job.count',
```

File: services/blockchain-indexer/config.js [hunk 14]
```diff
@@ -143,4 +143,28 @@ config.networks = Object.freeze({
...
+	},
+	refreshValidators: {
+		interval: process.env.REFRESH_VALIDATORS_INTERVAL || 0,
+		schedule: process.env.REFRESH_VALIDATORS_SCHEDULE || '*/5 * * * *',
+	},
+	validateValidatorsRank: {
+		interval: process.env.VALIDATE_VALIDATORS_RANK_INTERVAL || 0,
+		schedule: process.env.VALIDATE_VALIDATORS_RANK_SCHEDULE || '4-59/15 * * * *',
+	},
+	refreshIndexStatus: {
```
Comment (sameersubudhi) [inline services/blockchain-indexer/config.js:160]:
```suggestion
	refreshLiveIndexingJobCount: {
```

File: services/blockchain-indexer/jobs/dataService/indexStatus.js [hunk 15]
```diff
@@ -26,7 +27,8 @@ module.exports = [
...
 		description: 'Update the current number of jobs for block indexing in progress/wait.',
-		interval: 10, // Every 10 secs
+		interval: config.job.refreshIndexStatus.interval,
+		schedule: config.job.refreshIndexStatus.schedule,
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/indexStatus.js:31]:
```suggestion
		interval: config.job.refreshLiveIndexingJobCount.interval,
		schedule: config.job.refreshLiveIndexingJobCount.schedule,
```

File: services/blockchain-indexer/config.js [hunk 16]
```diff
@@ -143,4 +143,28 @@ config.networks = Object.freeze({
...
+		interval: process.env.REFRESH_INDEX_STATUS_INTERVAL || 10,
+		schedule: process.env.REFRESH_INDEX_STATUS_SCHEDULE || '* * * * *',
```
Comment (sameersubudhi) [inline services/blockchain-indexer/config.js:162]:
```suggestion
		schedule: process.env.REFRESH_INDEX_STATUS_SCHEDULE || '',
```

File: services/blockchain-indexer/jobs/dataService/validators.js [hunk 17]
```diff
@@ -15,16 +15,19 @@
...
+const config = require('../../config');
 const {
 	reloadValidatorCache,
 	isPosModuleRegistered,
 } = require('../../shared/dataService');
+const { validateValidatorCache } = require('../../shared/jobs/validateValidatorsCache');
 
 module.exports = [
 	{
 		name: 'reload.validators',
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/validators.js:27]:
```suggestion
		name: 'refresh.validators',
```

File: services/blockchain-indexer/jobs/dataService/validators.js [hunk 18]
```diff
@@ -47,4 +50,20 @@ module.exports = [
...
+		description: 'Validate cached validators and reload cache if necessary.',
+		interval: config.job.validateValidatorsRank.interval,
+		schedule: config.job.validateValidatorsRank.schedule,
+		controller: async () => {
+			if (await isPosModuleRegistered()) {
+				logger.debug('Validating validators cache...');
+				try {
+					await validateValidatorCache();
+				} catch (err) {
+					logger.warn(`Validating validators cache failed due to: ${err.message}`);
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/dataService/validators.js:64]:
```suggestion
					logger.warn(`Validating validators cache failed due to: ${err.message}.`);
```

File: services/blockchain-indexer/config.js [hunk 19]
```diff
@@ -143,4 +143,28 @@ config.networks = Object.freeze({
 
 config.db.isPersistEvents = Boolean(String(process.env.ENABLE_PERSIST_EVENTS).toLowerCase() === 'true');
 
+config.job = {
+	// Interval takes priority over schedule and must be greater than 0 to be valid
+	deleteSerializedEevents: {
```
Comment (sameersubudhi) [inline services/blockchain-indexer/config.js:148]:
```suggestion
	deleteSerializedEvents: {
```

File: services/blockchain-indexer/jobs/indexer/events.js [hunk 20]
```diff
@@ -23,7 +23,8 @@ module.exports = [
 	{
 		name: 'delete.serializedEvents',
 		description: 'Delete the serialized events until finalized height.',
-		schedule: '*/5 * * * *', // Every 5 min
+		interval: config.job.deleteSerializedEevents.interval,
+		schedule: config.job.deleteSerializedEevents.schedule,
```
Comment (sameersubudhi) [inline services/blockchain-indexer/jobs/indexer/events.js:27]:
```suggestion
		interval: config.job.deleteSerializedEvents.interval,
		schedule: config.job.deleteSerializedEvents.schedule,
```

File: services/gateway/jobs/ready.js [hunk 21]
```diff
@@ -13,13 +13,15 @@
  * Removal or modification of this copyright notice is prohibited.
  *
  */
+const config = require('../config');
 const { updateSvcStatus } = require('../shared/ready');
 
 module.exports = [
 	{
 		name: 'readiness.status',
```
Comment (sameersubudhi) [inline services/gateway/jobs/ready.js:21]:
```suggestion
		name: 'update.readiness.status',
```

File: services/market/jobs/pricesUpdate.js [hunk 22]
```diff
@@ -14,13 +14,16 @@
  *
  */
 const logger = require('lisk-service-framework').Logger();
+
+const config = require('../config');
 const { updatePrices } = require('../shared/market/priceUpdater');
 
 module.exports = [
 	{
 		name: 'prices.update',
```
Comment (sameersubudhi) [inline services/market/jobs/pricesUpdate.js:23]:
```suggestion
		name: 'update.prices',
```

File: services/market/config.js [hunk 23]
```diff
@@ -97,4 +97,28 @@ config.market.sources = {
...
+		schedule: process.env.REFRESH_BITTREX_PRICES_SCHEDULE || '* * * * *',
+	},
+	refreshExchangeratesapiPrices: {
+		interval: process.env.REFRESH_EXCHANGERATESAPI_PRICES_INTERVAL || 0,
+		schedule: process.env.REFRESH_EXCHANGERATESAPI_PRICES_SCHEDULE || '* * * * *',
+	},
+	refreshKrakenPrices: {
+		interval: process.env.REFRESH_KRAKEN_PRICES_INTERVAL || 0,
+		schedule: process.env.REFRESH_KRAKEN_PRICES_SCHEDULE || '* * * * *',
+	},
```
Comment (sameersubudhi) [inline services/market/config.js:117]:
```suggestion
	refreshPricesBinance: {
		interval: process.env.REFRESH_PRICES_BINANCE_INTERVAL || 0,
		schedule: process.env.REFRESH_PRICES_BINANCE_SCHEDULE || '* * * * *',
	},
	refreshPricesBittrex: {
		interval: process.env.REFRESH_PRICES_BITTREX_INTERVAL || 0,
		schedule: process.env.REFRESH_PRICES_BITTREX_SCHEDULE || '* * * * *',
	},
	refreshPricesExchangeratesapi: {
		interval: process.env.REFRESH_PRICES_EXCHANGERATESAPI_INTERVAL || 0,
		schedule: process.env.REFRESH_PRICES_EXCHANGERATESAPI_SCHEDULE || '* * * * *',
	},
	refreshPricesKraken: {
		interval: process.env.REFRESH_PRICES_KRAKEN_INTERVAL || 0,
		schedule: process.env.REFRESH_PRICES_KRAKEN_SCHEDULE || '* * * * *',
	},
```

File: services/market/jobs/binance.js [hunk 24]
```diff
@@ -35,7 +36,8 @@ module.exports = [
 	{
 		name: 'prices.retrieve.binance',
 		description: 'Fetches up-to-date market prices from Binance',
-		schedule: '* * * * *',
+		interval: config.job.refreshBinancePrices.interval,
+		schedule: config.job.refreshBinancePrices.schedule,
```
Comment (sameersubudhi) [inline services/market/jobs/binance.js:40]:
```suggestion
		interval: config.job.refreshPricesBinance.interval,
		schedule: config.job.refreshPricesBinance.schedule,
```

File: services/market/jobs/bittrex.js [hunk 25]
```diff
@@ -35,7 +36,8 @@ module.exports = [
 	{
 		name: 'prices.retrieve.bittrex',
 		description: 'Fetches up-to-date market prices from Bittrex',
-		schedule: '* * * * *',
+		interval: config.job.refreshBittrexPrices.interval,
+		schedule: config.job.refreshBittrexPrices.schedule,
```
Comment (sameersubudhi) [inline services/market/jobs/bittrex.js:40]:
```suggestion
		interval: config.job.refreshPricesBittrex.interval,
		schedule: config.job.refreshPricesBittrex.schedule,
```

File: services/market/jobs/kraken.js [hunk 26]
```diff
@@ -35,7 +36,8 @@ module.exports = [
 	{
 		name: 'prices.retrieve.kraken',
 		description: 'Fetches up-to-date market prices from Kraken',
-		schedule: '* * * * *',
+		interval: config.job.refreshKrakenPrices.interval,
+		schedule: config.job.refreshKrakenPrices.schedule,
```
Comment (sameersubudhi) [inline services/market/jobs/kraken.js:40]:
```suggestion
		interval: config.job.refreshPricesKraken.interval,
		schedule: config.job.refreshPricesKraken.schedule,
```

**Review by sameersubudhi — APPROVED** [review]
LGTM overall. Just a few minor recommendations below.

File: services/market/config.js [hunk 27]
```diff
@@ -97,4 +97,28 @@ config.market.sources = {
...
+	updatePrices: {
+		interval: process.env.UPDATE_PRICES_INTERVAL || 5,
+		schedule: process.env.UPDATE_PRICES_SCHEDULE || '* * * * *',
```
Comment (sameersubudhi) [inline services/market/config.js:120]:
```suggestion
		schedule: process.env.UPDATE_PRICES_SCHEDULE || '',
```

**Review by nagdahimanshu — APPROVED** [review]
LGTM 👍

**Pull request merged**